lib/Backend/i386/LowererMDArch.cpp

```diff
@@ -960,6 +960,45 @@ LowererMDArch::LowerAsmJsCallI(IR::Instr * callInstr)
     }
     return retInstr;
 }
+
+IR::Instr *
+LowererMDArch::LowerWasmMemOp(IR::Instr * instr, IR::Opnd *addrOpnd)
+{
+    IR::LabelInstr * helperLabel = Lowerer::InsertLabel(true, instr);
+    IR::LabelInstr * loadLabel = Lowerer::InsertLabel(false, instr);
+    IR::LabelInstr * doneLabel = Lowerer::InsertLabel(false, instr);
+
+    // Find array buffer length
+    IR::RegOpnd * indexOpnd = addrOpnd->AsIndirOpnd()->GetIndexOpnd();
+    Int64RegPair indexPair = lowererMD->m_lowerer->FindOrCreateInt64Pair(indexOpnd);
+    IR::Opnd *arrayLenOpnd = instr->GetSrc2();
+
+    // Compare index + memop access length and array buffer length, and generate RuntimeError if greater
+    IR::Opnd *cmpOpnd = indexPair.low;
+    // check high bits are zero
+    lowererMD->m_lowerer->InsertCompareBranch(indexPair.high, IR::IntConstOpnd::New(0, TyInt32, m_func), Js::OpCode::BrNeq_A, true, helperLabel, helperLabel);
+    // check low bits is within heapsize
+    lowererMD->m_lowerer->InsertCompareBranch(indexPair.low, arrayLenOpnd, Js::OpCode::BrGe_A, true, helperLabel, helperLabel);
+
+    if (addrOpnd->GetSize() > 1)
+    {
+        cmpOpnd = IR::RegOpnd::New(TyInt32, m_func);
+        // check low bits + size does not overflow
+        Lowerer::InsertAdd(true, cmpOpnd, indexPair.low, IR::IntConstOpnd::New(addrOpnd->GetSize() - 1, TyInt32, m_func), helperLabel);
+        Lowerer::InsertBranch(Js::OpCode::JO, helperLabel, helperLabel);
+        // check low bits + size is within heapsize
+        lowererMD->m_lowerer->InsertCompareBranch(cmpOpnd, arrayLenOpnd, Js::OpCode::BrGe_A, true, helperLabel, helperLabel);
+    }
+
+    // MGTODO : call RuntimeError once implemented
+    lowererMD->m_lowerer->GenerateRuntimeError(loadLabel, JSERR_InvalidTypedArrayIndex, IR::HelperOp_RuntimeRangeError);
+    Lowerer::InsertBranch(Js::OpCode::Br, loadLabel, helperLabel);
+
+    addrOpnd->AsIndirOpnd()->SetIndexOpnd(indexPair.low->AsRegOpnd());
```

Comment:
Right now it is true that `indexPair.low` is a RegOpnd, but once we implement const folding for int64, the index could be a constant opnd.
Since we have unit tests hitting that path, just put an assert for now and it will be easier to fix once we get there.